src/main/java/com/iota/iri/service/spentaddresses/impl/SpentAddressesServiceImpl.java

```diff
@@ -140,4 +124,113 @@ private boolean isBundleValid(Hash tailHash) throws Exception {
                 BundleValidator.validate(tangle, snapshotProvider.getInitialSnapshot(), tailHash);
         return (CollectionUtils.isNotEmpty(validation) && validation.get(0).get(0).getValidity() == 1);
     }
+
+    /**
+     *
+     * @param addressHash the address in question
+     * @param checkedAddresses a list of known unspent addresses, used to skip calculations
+     * @return {@code true} if address was spent from, else {@code false}
+     * @throws SpentAddressesException
+     * @see #wasAddressSpentFrom(Hash)
+     */
+    private boolean wasAddressSpentFrom(Hash addressHash, Set<Hash> checkedAddresses) throws SpentAddressesException {
+        if (addressHash == null) {
+            return false;
+        }
+
+        if (spentAddressesProvider.containsAddress(addressHash)) {
+            return true;
+        }
+
+        //If address has already been checked this session, return false
+        if (SafeUtils.isContaining(checkedAddresses, addressHash)){
+            return false;
+        }
+
+        //If address is either null or equal to the coordinator address
+        if (addressHash.equals(Hash.NULL_HASH) ||
```

Comment:
instead of adding 2 more checks here, `999` & `COO` can be added to `checkedAddress` initially.

it would make the code more readable (less pre checks) and reduce to equal ops.

Reply:
In order to nicely implement that suggestion I will need to solve #1315 (so the Set will be always from hashes). I can also create construct the Hash using the factory but let's do it right already

Reply:
In the end #1315 is blocked so I used `HashFactory`